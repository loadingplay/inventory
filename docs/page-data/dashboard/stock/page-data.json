{"componentChunkName":"component---src-pages-dashboard-stock-index-jsx","path":"/dashboard/stock/","webpackCompilationHash":"2779579faed300d22aca","result":{"pageContext":{"isCreatedByStatefulCreatePages":true}}}