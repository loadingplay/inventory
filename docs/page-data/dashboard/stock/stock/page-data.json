{"componentChunkName":"component---src-pages-dashboard-stock-stock-jsx","path":"/dashboard/stock/stock/","webpackCompilationHash":"2779579faed300d22aca","result":{"pageContext":{"isCreatedByStatefulCreatePages":true}}}