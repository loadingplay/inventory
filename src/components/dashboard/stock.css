.stock-message {
    height: 4600px;
}
