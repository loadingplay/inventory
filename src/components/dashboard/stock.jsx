import React, { useEffect, useState } from 'react'
import ReactDOM from 'react-dom';
import Paginator from 'react-hooks-paginator';
import './stores.css'
import authService from '../../services/authService';
import stockService from '../../services/stockService';

const stock = (props) => {
  const [data, dataSet] = useState([]);
  const [offset, setOffset] = useState(0);
  const [currentPage, setCurrentPage] = useState(1);
  const [currentData, setCurrentData] = useState([]);
  const token = authService.getCurrentToken();
  const pageLimit = 10;

  async function fetchSuggestedData() {
    let stocksSuggested = await stockService.getSuggestedProducts(props.cellarId, token);
    let data = stocksSuggested.data.replenishments
    data ? data.sort((a, b) => (a.suggested > b.suggested) ? 1 : -1) : '';
    dataSet(data);
  }

  async function fetchStockData(){
    let getStock = await stockService.getStockByCellarId(props.cellarId, token)
    let dataProducts = getStock.data.products;
    dataProducts ? dataProducts.map((item, i) => {
      currentData ? currentData.map((e, i) => {
        if(e.sku === item.product_sku){
          e.balance_units = item.balance_units
          e.in_stock = item.in_stock
        }
        return e;
      }) : '';
      return currentData;
    }) : '';
  }

  useEffect(() => {
    fetchSuggestedData();
  }, [props.cellarId])

  useEffect(() => {
    fetchStockData();
  }, [props.cellarId])

  useEffect(() => {
    setCurrentData(data ? data.slice(offset, offset + pageLimit) : []);
    fetchStockData();
  }, [offset, data, props.cellarId]);

  async function handleSearch(evt) {
    let sku = evt.target.value;
    let currentInventory = await stockService.getCurrentInventory(props.cellarId, token, sku);
    console.log(currentInventory);
  }


  return (
    <section className="stores_wrapper">
      <table className="table table-hover table-borderless">
        <thead>
          <tr className="table-info">
            <th scope="col">sku</th>
            <th scope="col">nombre producto</th>
            <th scope="col">stock en bodega</th>
            <th scope="col">stock en tienda</th>
            <th scope="col">sugerido</th>
            <th scope="col">picker</th>
          </tr>
        </thead>
        <tbody>
          {currentData ? currentData.map((item, i) =>
            <tr key={i}>
              <th scope="row">{item.sku}</th>
              <td>{item.site_name}</td>
              <td>{item.in_stock ? 'Sí' : 'No'}</td>
              <td>{item.balance_units ? item.balance_units : 0}</td>
              <td>{item.suggested}</td>
              <td></td>
            </tr>
          ) : <tr><td>Buscando...</td></tr>}
        </tbody>
      </table>
      <Paginator
        totalRecords={data ? data.length : 1}
        pageLimit={4}
        pageNeighbours={1}
        setOffset={setOffset}
        currentPage={currentPage}
        setCurrentPage={setCurrentPage}
      />
      <section className="search_actions">
        <div>
          <input type="text" placeholder="ingrese sku" onChange={handleSearch} />
        </div>
        <div className="list_btn">
          <div>
            <button type="button" className="btn btn-success">
              imprimir
              </button>
          </div>
          <div>
            <button type="button" className="btn btn-primary">
              generar guia de despacho
              </button>
          </div>
          <div>
            <button type="button" className="btn btn-danger" onClick={authService.logout}>
              cerrar sesión
              </button>
          </div>
        </div>
      </section>
    </section>
  )
}

export default stock;