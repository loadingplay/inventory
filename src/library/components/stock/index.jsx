export * from "./stock";
export * from "../picker/PickerControl";
