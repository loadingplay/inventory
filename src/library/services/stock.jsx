var Loader = function (access_token) {
  this.access_token = access_token;
  this.instances = 0;
  this.max_instances = 15;
  this.chunk_size = 10;
};

Loader.prototype.load = function (cellar_id_list, sku_list) {
  var loader = this;
  cellar_id_list.forEach(function (cellar) {
    loader.loadCellar(cellar, sku_list);
  });
  return this;
};

Loader.prototype.loadCellar = function (cellarid, sku_list) {
  var i, j, temparray;
  for (i = 0, j = sku_list.length; i < j; i += this.chunk_size) {
    temparray = sku_list.slice(i, i + this.chunk_size);
    // do whatever
    this.performRequest(cellarid, temparray);
  }
};

Loader.prototype.performRequest = function (cellarid, skus_chunk) {
  var loader = this;
  if (loader.instances > loader.max_instances) {
    setTimeout(function () {
      loader.performRequest(cellarid, skus_chunk);
    }, 100);
    return
  }
  this.instances += 1;
  fetch("https://apibodegas.loadingplay.com/v1/cellar/" + cellarid + "/product/?sku_list=" + skus_chunk.join(","),
  {
    headers: {
      "Authorization": "Bearer " + this.access_token
    }
  })
  .then((response) => response.json())
  .then((response) => {
    loader.instances -= 1;
    loader.callback(cellarid, response.products);
  });
};


Loader.prototype.done = function (callback) {
  this.callback = callback
};

export function StockLoader(access_token) {
  return new Loader(access_token);
}
