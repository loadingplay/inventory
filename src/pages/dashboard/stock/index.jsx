import "./stock.css";
export * from "./stock";
export * from "../../../library/components/picker/PickerControl"
