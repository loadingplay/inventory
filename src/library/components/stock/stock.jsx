import React, { Component } from 'react';

import { StockLoader, StoreLoader } from "../../services";
import { PickerControl } from '..';
import "./stock.css";

// implement stock class
export class Stock extends Component {

  constructor(props) {
    super(props);
    this.state = {
      products: null,
      error_message: ""
    };
  }

  handleNewStatus = (json_data) => {
    // handle status
    if (json_data.status === "error") {
      this.setState({
        error_message: json_data.message
      });
    }
    else
    {
      this.setState({
        products: json_data.replenishments,
        error_message: ""
      });
    }
  }

  loadInventory = async (cellar_id, page) => {
    let store_loader,
      json_data;

    // reset status
    this.setState({ products: null });
    store_loader = new StoreLoader(this.props.accessToken);
    json_data = await store_loader.loadProducts(page, cellar_id);

    this.handleNewStatus(json_data);

    // load current inventory
    if (json_data.replenishments)
      this.loadStoreInventory(cellar_id, json_data.replenishments);
  }

  loadStoreInventory = (cellar_id, products) => {
    // early termination
    if (products === undefined) return;

    let skus = products.map((item) => {
      return item.sku;
    });

    StockLoader(this.props.accessToken)
    .load([cellar_id], skus)
    .done((cellar_id, products) => {
      if (!this.state.products) return;
      let new_products = this.state.products.map((item, index) => {
        let inventory = products.find((element) => {
          return item.sku === element.product_sku;
        });
        if (item.current_inventory === undefined) {
          item.current_inventory = inventory !== undefined ? inventory.balance_units:undefined;
        }
        return item;
      });

      this.setState({
        products: new_products
      });
    });

  }

  componentWillReceiveProps = (newProps) => {
    if (newProps.selectedCellar !== 0) {
      this.loadInventory(newProps.selectedCellar, newProps.currentPage + 1);
    }
  }

  renderProductList = () => {
    let products;

    if (!this.props.selectedCellar)
    {
      products = (<tr><td colSpan="6" className="stock-message" >Seleccione una bodega</td></tr>);
    }
    else if (this.state.error_message !== "")
    {
      products = (<tr><td colSpan="6" className="stock-message" >{this.state.error_message}</td></tr>);
    }
    else if (!this.state.products)
    {
      products = (<tr><td colSpan="6" className="stock-message" >Cargando...</td></tr>);
    }
    else if (this.state.products.length === 0)
    {
      products = (<tr><td colSpan="6" >No hay productos sugeridos en esta tienda</td></tr>);
    }
    else
    {
      products = this.state.products.map((item, index) => {
        return (
          <tr key={index} >
            <th scope="row">{item.sku}</th>
            <td></td>
            <td></td>
            <td>
              {
                item.current_inventory === undefined ?
                'cargando...':item.current_inventory
              }
            </td>
            <td>{item.suggested}</td>
            <td><PickerControl cellar_id={this.props.selectedCellar} sku={item.sku} ></PickerControl></td>
          </tr>
        );
      });
    }

    return products;
  }

  render() {

    return (
      <section className="stores_wrapper col-12">
        <table className="table table-sm">
          <thead>
            <tr className="table-info">
              <th scope="col">sku</th>
              <th scope="col">nombre</th>
              <th scope="col">stock bodega</th>
              <th scope="col">stock tienda</th>
              <th scope="col">sugerido</th>
              <th scope="col">picker</th>
            </tr>
          </thead>
          <tbody>
            { this.renderProductList() }
          </tbody>
        </table>
      </section>
    )
  }
}
