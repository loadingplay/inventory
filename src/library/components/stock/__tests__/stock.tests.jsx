import React from 'react';
import { shallow } from "enzyme";

import { Stock } from "../stock";

describe("Stock", () => {
  test("it should render empty", () => {
    const wrapper = shallow(<Stock></Stock>);
    expect(wrapper).toMatchSnapshot("empty");
  });

  test("it should render with loading message", () => {
    const wrapper = shallow(
      <Stock
        selectedCellar={1}
        hasError={false}
        errorMessage=""
        isLoading={true}
        extraProducts={[]}
        products={[]}
        loadKey={1}
      ></Stock>
    );
    expect(wrapper).toMatchSnapshot("loading message");
  });

  test("it should render with error message", () => {
    const wrapper = shallow(
      <Stock
        selectedCellar={1}
        hasError={true}
        errorMessage="error message"
        isLoading={false}
        extraProducts={[]}
        products={[]}
        loadKey={1}
      ></Stock>
    );
    expect(wrapper).toMatchSnapshot("error message");
  });

  test("it should render with products", () => {
    const wrapper = shallow(
      <Stock
        selectedCellar={1}
        hasError={false}
        errorMessage=""
        isLoading={false}
        extraProducts={[]}
        products={[{}]}
        loadKey={1}
      />
    );
    expect(wrapper).toMatchSnapshot("with products");
  });
});
