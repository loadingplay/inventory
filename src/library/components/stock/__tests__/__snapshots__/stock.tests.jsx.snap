// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stock it should render empty: empty 1`] = `
<section
  className="stores_wrapper col-12"
>
  <table
    className="table table-sm"
  >
    <thead>
      <tr
        className="table-info"
      >
        <th
          scope="col"
        >
          sku
        </th>
        <th
          scope="col"
        >
          nombre
        </th>
        <th
          scope="col"
        >
          stock bodega
        </th>
        <th
          scope="col"
        >
          stock tienda
        </th>
        <th
          scope="col"
        >
          sugerido
        </th>
        <th
          scope="col"
        >
          picker
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</section>
`;

exports[`Stock it should render with error message: error message 1`] = `
<section
  className="stores_wrapper col-12"
>
  <table
    className="table table-sm"
  >
    <thead>
      <tr
        className="table-info"
      >
        <th
          scope="col"
        >
          sku
        </th>
        <th
          scope="col"
        >
          nombre
        </th>
        <th
          scope="col"
        >
          stock bodega
        </th>
        <th
          scope="col"
        >
          stock tienda
        </th>
        <th
          scope="col"
        >
          sugerido
        </th>
        <th
          scope="col"
        >
          picker
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="stock-message"
          colSpan="6"
        >
          error message
        </td>
      </tr>
    </tbody>
  </table>
</section>
`;

exports[`Stock it should render with loading message: loading message 1`] = `
<section
  className="stores_wrapper col-12"
>
  <table
    className="table table-sm"
  >
    <thead>
      <tr
        className="table-info"
      >
        <th
          scope="col"
        >
          sku
        </th>
        <th
          scope="col"
        >
          nombre
        </th>
        <th
          scope="col"
        >
          stock bodega
        </th>
        <th
          scope="col"
        >
          stock tienda
        </th>
        <th
          scope="col"
        >
          sugerido
        </th>
        <th
          scope="col"
        >
          picker
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="stock-message"
          colSpan="6"
        >
          Cargando...
        </td>
      </tr>
    </tbody>
  </table>
</section>
`;

exports[`Stock it should render with products: with products 1`] = `
<section
  className="stores_wrapper col-12"
>
  <table
    className="table table-sm"
  >
    <thead>
      <tr
        className="table-info"
      >
        <th
          scope="col"
        >
          sku
        </th>
        <th
          scope="col"
        >
          nombre
        </th>
        <th
          scope="col"
        >
          stock bodega
        </th>
        <th
          scope="col"
        >
          stock tienda
        </th>
        <th
          scope="col"
        >
          sugerido
        </th>
        <th
          scope="col"
        >
          picker
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="0"
      >
        <th
          scope="row"
        />
        <td />
        <td>
          cargando...
        </td>
        <td>
          cargando...
        </td>
        <td />
        <td>
          <PickerControl
            cellar_id="1"
            item={Object {}}
            key="1"
          />
        </td>
      </tr>
    </tbody>
  </table>
</section>
`;
